sipa/base.py

```diff
@@ -41,11 +41,11 @@ def babel_selector():
 
     if 'locale' in request.args and Locale(
             request.args['locale']) in possible_locales():
-        session['locale'] = request.args['locale']
-    elif not session.get('locale'):
+        return request.args['locale']
+    elif request.cookies.get("lang") is None:
         langs = []
         for lang in possible_locales():
             langs.append(lang.language)
-        session['locale'] = request.accept_languages.best_match(langs)
+            return request.accept_languages.best_match(langs)
 
-    return session.get('locale')
+    return request.cookies.get("lang", "en")
```

Comment:
Also, building up a list from another is a great use case for a [list comprehension](https://docs.python.org/3.5/tutorial/datastructures.html#list-comprehensions):

```python
languages = [locale.language for locale in possible_locales()]
return request.accept_languages.best_match(langs)
```

Even better, put this directly into the function, since it is only used there (plus, use a genexp):
```python
return  request.accept_languages.best_match(locale.language for locale in possible_locales)
```